optuna/integration/chainermn.py

```diff
@@ -168,66 +167,67 @@ def __setattr__(self, attr_name, value):
-                self.comm.mpi_comm.bcast(result)
-                return result
-            except Exception as e:
-                self.comm.mpi_comm.bcast(e)
-                raise
-        else:
-            result = self.comm.mpi_comm.bcast(None)
-            if isinstance(result, Exception):
-                raise result
-            return result
-
-
-class _ChainerMNTrial(Trial):
+class ChainerMNTrial(BaseTrial):
     def __init__(self, trial, comm):
-        # type: (Trial, CommunicatorBase) -> None
+        # type: (Optional[Trial], CommunicatorBase) -> None
 
-        super(_ChainerMNTrial, self).__init__(trial.study,
-                                              trial._trial_id,
-                                              disable_relative_sampling=True)
         self.delegate = trial
         self.comm = comm
 
     def suggest_uniform(self, name, low, high):
         # type: (str, float, float) -> float
 
-        return self._suggest_with_mpi(self.delegate.suggest_uniform, *(name, low, high))
+        def func():
+            # type: () -> float
+
+            assert self.delegate is not None
+            return self.delegate.suggest_uniform(name, low, high)
+
+        return self._call_with_mpi(func)
 
     def suggest_loguniform(self, name, low, high):
         # type: (str, float, float) -> float
 
-        return self._suggest_with_mpi(self.delegate.suggest_loguniform, *(name, low, high))
+        def func():
+            # type: () -> float
+
+            assert self.delegate is not None
+            return self.delegate.suggest_loguniform(name, low, high)
+
+        return self._call_with_mpi(func)
 
     def suggest_discrete_uniform(self, name, low, high, q):
         # type: (str, float, float, float) -> float
 
-        return self._suggest_with_mpi(self.delegate.suggest_discrete_uniform,
-                                      *(name, low, high, q))
+        def func():
+            # type: () -> float
+
+            assert self.delegate is not None
+            return self.delegate.suggest_discrete_uniform(name, low, high, q)
+
+        return self._call_with_mpi(func)
 
     def suggest_int(self, name, low, high):
         # type: (str, int, int) -> int
 
-        return self._suggest_with_mpi(self.delegate.suggest_int, *(name, low, high))
+        def func():
+            # type: () -> int
+
+            assert self.delegate is not None
+            return self.delegate.suggest_int(name, low, high)
+
+        return self._call_with_mpi(func)
 
     def suggest_categorical(self, name, choices):
         # type: (str, Sequence[T]) -> T
 
-        return self._suggest_with_mpi(self.delegate.suggest_categorical, *(name, choices))
+        def func():
+            # type: () -> T
+
+            assert self.delegate is not None
+            return self.delegate.suggest_categorical(name, choices)
+
+        return self._call_with_mpi(func)
 
     def _suggest_with_mpi(self, func, *args):
```

Comment:
`_suggest_with_mpi` is not called from anywhere, and I think we can remove it.

Reply:
Oops..., thank you for pointing it out!

Fixed: https://github.com/pfnet/optuna/pull/427/commits/3e4cef8f4a95c1c22cc34548e7be3f62a38dcd8f